### ISSUE TITLE
Improve scoring algorithm

### ISSUE THREAD

**Issue by Chicoryn** [Chicoryn/dream-go#26]
Our scoring algorithm is pretty bad at detecting large dead dragons, seki, and life and death in general. I suggest we solve this as it affects the _value head_ greatly (since we train it to recognize the wrong result)

---

Based on 1,000 policy play games the current score is wrong 3% of the time. This could be improved, I suggest we aim for <1%, any lower number is likely unrealistic as it is going to involve very unfinished situations:

```
kalle@coffee:[dream-go-copy]$ cargo run --release -- --policy-play 1000 > policy_play.sgf
kalle@coffee:[dream-go-copy]$ python3 ../dream-go/tools/sgf2score.py < policy_play.sgf > policy_play_wrong.sgf
Arbitrated 0 games without a winner
Arbitrated 30 games with the wrong winner
```

The `sgf2score.py` script uses GNU Go with Chinese scoring to determine the winner, and compare it against what is recorded in the SGF file.

**Comment by Chicoryn** [comment]
I adjusted the algorithm to use a simple rule based policy instead of the flood filling technique previously used:

1. Make territories as safe as possible, by playing moves that result in a net-gain of liberties either by extending into one of our liberties or capturing opponent groups that are in atari.
2. Try to capture opponent groups by filling its liberties in a _safe_ manner.
3. Flood-fill to determine the life & death of any remaining groups (since some groups may remain after step 2 if there are no _safe_ moves to capture a group).

This works pretty well, I also experimented with a much more aggressive policy but it ended up being too complicated as it tried to solve capturing races, etc. The one listed above is **much** simpler and therefore does not solve all the provided scenarios but assuming the players has already done most of the heavy lifting in making sure groups are alive it works very well:

```
kalle@coffee:[dream-go]$ ./dream_go --policy-play 1000 > policy_play.sgf 
kalle@coffee:[dream-go]$ python tools/sgf2score.py < policy_play.sgf > policy_play_correct.sgf
Arbitrated 0 games without a winner
Arbitrated 15 games with the wrong winner
```

All of the provided games except 3 are either small point wins where GNU Go disagree (because we handle passes differently), unfinished positions that either player could win if they make the next move, or seki which the scoring algorithm is not great at solving yet. However, this is still 1.5% which is larger than the goal, so I will try to fix the games I consider bugs.

---

To examine the moves chosen by the policy you can run the following commands in the GTP prompt, the `sabaki-genmovelog` will generate an SGF file containing the moves. If you are using the [Sabaki](https://github.com/SabakiHQ/Sabaki) client then it will automatically be added to your game-tree:

```
final_score
= B+0.5
sabaki-genmovelog
= #sabaki{...}
```

**Comment by Chicoryn** [comment]
After fixing two of the bugs listed above, we now perform slightly better as the policy no longer:

- Play unnecessary self-atari moves (mostly useful to avoid _resolving_ some sekis).
- Prefer to capture over running away to gain liberties.

The remaining bug is related to big eyes, where the policy does not recognize the urgency in playing the final move, and capturing races:

```
kalle@coffee:[dream-go]$ ./dream_go --policy-play 1000 > policy_play.sgf 
kalle@coffee:[dream-go]$ python tools/sgf2score.py < policy_play.sgf > policy_play_correct.sgf
Arbitrated 0 games without a winner
Arbitrated 13 games with the wrong winner
```

**Comment by Chicoryn** [comment]
There are also a few issues during the first phase, where sometimes a group will extend their liberties _too much_ and end up capturing a group it _shouldn't_ (since it would not normally be able to live).

I am unsure about how to fix this issue, might be fine since I've only noticed it occurring twice over 4,000 games.

---

If I want to go even further down the rabbit hole; I am using the following implementation in GNU Go as an oracle, and if I think it is that good, then why not re-implement it in rust?

https://github.com/liangsirrun/gnugo/blob/master/engine/aftermath.c

**Comment by Chicoryn** [comment]
The current algorithm seems to fit our bill of less than 1% error ratio in practice, so I am going to close this issue.

In addition to that, I changed the worker image to re-score all self-play games using GNU Go (which does very well once the game is finished, even if I would not trust it to finish the game) so it should now recognize seki, and other weird cases correctly during self play.
